### PR TITLE
Add Domino online training path

### PR DESCRIPTION
**Pull request by jianuo-huang**
## 1. Objective

Implement Domino online training in SpecForge.

Domino extends the parallel DFlash draft backbone with a lightweight causal correction
head. The backbone computes base logits for a full draft block in parallel, and the
Domino head refines those logits with a GRU state built from previous draft tokens.

- Input: Target-model context hidden states from the online DFlash target backend.
- Output: A Domino draft model checkpoint compatible with the existing DFlash draft
  model format plus Domino head parameters.
- Mode: Online training with live target-model inference.

The implementation follows the Domino method described in https://arxiv.org/pdf/2605.29707.

### 2.1 Module Structure

| Component | Path | Responsibility |
|---|---|---|
| Training core | `specforge/core/domino.py` | `OnlineDominoModel`, block construction, Domino loss, metrics |
| Draft model | `specforge/modeling/draft/dflash.py` | Optional Domino head modules on top of `DFlashDraftModel` |
| Training script | `scripts/train_domino.py` | Domino online training entrypoint |
| Config | `configs/qwen3-8b-domino.json` | Qwen3-8B Domino draft config |
| Example | `examples/run_qwen3_8b_domino_online.sh` | Example Qwen3-8B online training command |

The existing DFlash training path is left unchanged. Domino uses a separate training
wrapper and script while reusing the DFlash target backend and draft backbone.

### 2.2 Integration Point

Domino reuses the existing DFlash online target backend and parallel draft backbone.
The new path starts after the backbone hidden states are produced: base logits are
computed with the frozen target LM head, and the Domino head adds a GRU-conditioned
low-rank correction before the CE loss.

### 3.1 Domino Head

For `projector_type="domino"`, `DFlashDraftModel` adds:

- A GRU causal encoder over previous draft-token embeddings.
- A low-rank logit correction head with bottleneck dimension `emb_dim`.
- A residual correction path that adds the Domino correction to the parallel base
  logits.

### 3.2 Training Objective

Domino jointly supervises:

- `final_loss`: cross entropy on Domino-refined logits.
- `base_loss`: cross entropy on backbone-only base logits.

The training objective uses the base-anchor curriculum:

```text
loss = (1 - lambda_base) * final_loss + lambda_base * base_loss
```

`lambda_base` linearly decays during training, so early training keeps the parallel
backbone strong and later training lets the Domino correction head take over.

### 3.3 Metrics

The training core reports:

- `accuracy` and `accept_len` for Domino-refined logits.
- `base_accuracy` and `base_accept_len` for backbone-only logits.
- `final_loss`, `base_loss`, and `lambda_base`.

These metrics are used for training diagnostics only and do not affect gradients.

## 4. Files Changed

```text
configs/qwen3-8b-domino.json
examples/run_qwen3_8b_domino_online.sh
scripts/train_domino.py
specforge/core/__init__.py
specforge/core/domino.py
specforge/modeling/draft/dflash.py
```

## 5. Usage

```bash
bash examples/run_qwen3_8b_domino_online.sh
```

The example expects the same target-model and dataset setup as the existing DFlash
online training examples, with Domino architecture parameters read from
`configs/qwen3-8b-domino.json`.

## 6. Validation

Local checks passed:

```bash
pre-commit run --all-files --show-diff-on-failure

python -m py_compile \
  specforge/core/dflash.py \
  specforge/core/domino.py \
  scripts/train_dflash.py \
  scripts/train_domino.py \
  specforge/modeling/draft/dflash.py

python -m json.tool configs/qwen3-8b-domino.json >/dev/null
bash -n examples/run_qwen3_8b_domino_online.sh
```

Targeted backend/JIT checks passed with `dLLM_env`, a persistent SGLang/JIT cache,
and an existing HF cache:

```bash
python -m unittest \
  tests.test_modeling.test_target.test_sglang_backend.test_sglang_backend.TestTargetModelBackend.test_sglang_backend_with_dense \
  -v

python -m unittest \
  tests.test_modeling.test_target.test_target_model_backend.TestTargetModelBackend \
  -v
```

| Check | Result |
|---|---:|
| `test_sglang_backend_with_dense` | OK, 1 test |
| `TestTargetModelBackend` | OK, 2 tests |

## 7. Training Results

I ran an end-to-end Qwen3-8B Domino training job on the ShareGPT same-data setting.
The run completed 6 epochs / 29286 training steps and produced the final checkpoint
successfully.

Qwen3-8B Domino training curves

The curves track the two supervised logits used by Domino training. `final_loss` is
computed from the Domino-refined logits, while `base_loss` is computed from the
parallel backbone logits before correction. `accuracy` and `accept_len` are reported
for the refined logits, with `base_accuracy` and `base_accept_len` showing the
backbone-only baseline. `lambda_base` is the base-anchor curriculum weight.

| Item | Value |
|---|---|
| Hardware | 8 x NVIDIA A100-SXM4-80GB |
| Runtime | 8h 48m 17s |
| Target model | Qwen3-8B |
| Dataset | ShareGPT regenerated data |
| Target backend | SGLang |
| Attention backend | `flex_attention` |
| Block size | 16 |
| Max length | 3072 |
| Number of anchors | 256 |
| Domino prefix length | 1 |
| Shift label | enabled |
| GRU hidden dim | 1024 |
| Bottleneck dim | 256 |
| Final checkpoint | epoch 6 / step 29286 |

Final training metrics reached `train/loss=0.9595`, `train/final_loss=0.9579`,
`train/base_loss=2.3111`, `train/accuracy=0.7180`, and `train/accept_len=6.9208`.

## 8. Downstream Evaluation

I also evaluated the final Domino checkpoint against the DFlash baseline across
reasoning, coding, and instruction-following benchmarks. The table reports average
accepted length from the existing benchmark pipeline; higher is better.
The DFlash baseline and Domino checkpoint are trained under the same data and
training setup, and evaluated with the same target model, benchmark harness, and
decoding configuration.

| Benchmark | DFlash accept len | Domino accept len | Relative gain |
|---|---:|---:|---:|
| GSM8K | 4.03 | 4.78 | +18.6% |
| MATH500 | 3.98 | 4.64 | +16.6% |
| AIME24 | 4.02 | 4.66 | +15.9% |
| AIME25 | 3.75 | 4.30 | +14.7% |
| HumanEval | 3.76 | 4.35 | +15.7% |
| MBPP | 3.39 | 3.94 | +16.2% |
| MT-Bench | 2.92 | 3.40 | +16.4% |

## 9. References

- Domino: https://arxiv.org/pdf/2605.29707
- Domino reference implementation: https://github.com/jianuo-huang/Domino

### CODE REVIEW

**Comment by jiapingW** [comment]
Great work!
